### PR TITLE
Add new sample for ray tracing position fetch

### LINKED ISSUE #1004: Add new sample for VK_KHR_ray_tracing_position_fetch
While it has limited uses, the recent VK_KHR_ray_tracing_position_fetch extension is well suited for a new sample. To raise awareness, I'll be doing a sample on this.

### PR DESCRIPTION
**Pull request by SaschaWillems**
## Description

This PR adds a new sample for the VK_KHR_ray_tracing_position_fetch extension. It comes with a small tutorial, and shaders both for GLSL and HLSL.

Developed and tested on an nvidia RTX 4070 and Windows 11.

**Important note:** Requires new assets from this PR: https://github.com/KhronosGroup/Vulkan-Samples-Assets/pull/26

Fixes #1004

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Copyright-Notice-and-License-Template)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

 Note: The Samples CI runs a number of checks including:
 - [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
 - [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)

 If this PR contains framework changes:
 - [ ] I did a full batch run using the `batch` command line argument to make sure all samples still work properly

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the readme of the folder that the sample belongs to [e.g. api samples readme](https://github.com/KhronosGroup/Vulkan-Samples/blob/main/samples/api/README.adoc)
- [x] For new samples, I have added a tutorial README.md file to guide users through what they need to know to implement code using this feature. For example, see [conditional_rendering](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/samples/extensions/conditional_rendering)
- [x] For new samples, I have added a link to the [Antora navigation](https://github.com/KhronosGroup/Vulkan-Samples/blob/main/antora/modules/ROOT/nav.adoc) so that the sample will be listed at the Vulkan documentation site

### CODE REVIEW

File: framework/api_vulkan_sample.cpp [hunk 1]
```diff
@@ -1308,11 +1308,11 @@ Texture ApiVulkanSample::load_texture_cubemap(const std::string &file, vkb::sg::
 	return texture;
 }
 
-std::unique_ptr<vkb::sg::SubMesh> ApiVulkanSample::load_model(const std::string &file, uint32_t index, bool storage_buffer)
+std::unique_ptr<vkb::sg::SubMesh> ApiVulkanSample::load_model(const std::string &file, uint32_t index, bool storage_buffer, VkBufferUsageFlags additional_buffer_usage_flags)
```
Comment (asuessenbach) [inline framework/api_vulkan_sample.cpp:1311]: Would you mind to port those changes to the HPPApiVulkanSample as well?

File: framework/core/acceleration_structure.cpp [hunk 2]
```diff
@@ -66,6 +66,32 @@ uint64_t AccelerationStructure::add_triangle_geometry(std::unique_ptr<vkb::core:
 	return index;
 }
 
+uint64_t AccelerationStructure::add_triangle_geometry(vkb::core::Buffer                  &vertex_buffer,
+                                                      std::unique_ptr<vkb::core::Buffer> &index_buffer,
+                                                      std::unique_ptr<vkb::core::Buffer> &transform_buffer,
```
Comment (asuessenbach) [inline framework/core/acceleration_structure.cpp]: Maybe change all three arguments, `vertex_buffer`, `index_buffer`, and `transform_buffer` to a `vkb::core::Buffer const&`?

File: framework/gltf_loader.cpp [hunk 3]
```diff
@@ -749,9 +749,15 @@ sg::Scene GLTFLoader::load_scene(int scene_index)
...
+				VkBufferUsageFlags buffer_usage_flags = VK_BUFFER_USAGE_VERTEX_BUFFER_BIT;
+				if (additional_buffer_usage_flags)
+				{
+					buffer_usage_flags |= additional_buffer_usage_flags;
+				}
+
 				core::Buffer buffer{device,
 				                    vertex_data.size(),
-				                    VK_BUFFER_USAGE_VERTEX_BUFFER_BIT,
+				                    buffer_usage_flags,
```
Comment (asuessenbach) [inline framework/gltf_loader.cpp]: No need to introduce this local variable, you can directly use `VK_BUFFER_USAGE_VERTEX_BUFFER_BIT | additional_buffer_usage_flags` here.

File: framework/gltf_loader.cpp [hunk 4]
```diff
@@ -792,9 +798,15 @@ sg::Scene GLTFLoader::load_scene(int scene_index)
...
+				VkBufferUsageFlags buffer_usage_flags = VK_BUFFER_USAGE_INDEX_BUFFER_BIT;
+				if (additional_buffer_usage_flags)
+				{
+					buffer_usage_flags |= additional_buffer_usage_flags;
+				}
+
 				submesh->index_buffer = std::make_unique<core::Buffer>(device,
 				                                                       index_data.size(),
-				                                                       VK_BUFFER_USAGE_INDEX_BUFFER_BIT,
+				                                                       additional_buffer_usage_flags,
```
Comment (asuessenbach) [inline framework/gltf_loader.cpp]: Did you want to use `buffer_usage_flags` here, that is, with `VK_BUFFER_USAGE_INDEX_BUFFER_BIT`?
If so, the comment from above holds here as well.

File: framework/gltf_loader.cpp [hunk 5]
```diff
@@ -1190,9 +1204,15 @@ std::unique_ptr<sg::SubMesh> GLTFLoader::load_model(uint32_t index, bool storage
...
+		VkBufferUsageFlags buffer_usage_flags = VK_BUFFER_USAGE_TRANSFER_DST_BIT | VK_BUFFER_USAGE_VERTEX_BUFFER_BIT;
+		if (additional_buffer_usage_flags)
+		{
+			buffer_usage_flags |= additional_buffer_usage_flags;
+		}
+
 		core::Buffer buffer{device,
 		                    vertex_data.size() * sizeof(Vertex),
-		                    VK_BUFFER_USAGE_TRANSFER_DST_BIT | VK_BUFFER_USAGE_VERTEX_BUFFER_BIT,
+		                    buffer_usage_flags,
```
Comment (asuessenbach) [inline framework/gltf_loader.cpp:1203]: As above, no need for the local variable `buffer_usage_flags`.

File: framework/gltf_loader.cpp [hunk 6]
```diff
@@ -1260,9 +1280,15 @@ std::unique_ptr<sg::SubMesh> GLTFLoader::load_model(uint32_t index, bool storage
...
+			VkBufferUsageFlags buffer_usage_flags = VK_BUFFER_USAGE_TRANSFER_DST_BIT | VK_BUFFER_USAGE_INDEX_BUFFER_BIT;
+			if (additional_buffer_usage_flags)
+			{
+				buffer_usage_flags |= additional_buffer_usage_flags;
+			}
+
 			submesh->index_buffer = std::make_unique<core::Buffer>(device,
 			                                                       index_data.size(),
-			                                                       VK_BUFFER_USAGE_TRANSFER_DST_BIT | VK_BUFFER_USAGE_INDEX_BUFFER_BIT,
+			                                                       buffer_usage_flags,
```
Comment (asuessenbach) [inline framework/gltf_loader.cpp:1279]: As above.

File: samples/extensions/README.adoc [hunk 7]
```diff
@@ -173,6 +173,12 @@ Then it uses that dynamic  sample rate map as a base for the next frame.
 Render a simple scene showing the basics of ray tracing, including reflection and shadow rays.
 The sample creates some geometries and create a bottom acceleration structure for each, then make instances of those, using different materials and placing them at different locations.
 
+=== xref:./{extension_samplespath}ray_tracing_position_fetch/README.adoc[Ray tracing position fetch]
+
+*Extensions*: https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VK_KHR_ray_tracing_position_fetch[`VK_KHR_ray_tracing_position_fetch`]
+
+Shows how to use the ray tracing position fetch extension to directly access vertex positions for a hit triangle from the acceleration strucuture, instead of having to explicitly pass and unpack that information
```
Comment (asuessenbach) [inline samples/extensions/README.adoc]: Typo: strucuture -> structure

File: samples/extensions/ray_tracing_position_fetch/ray_tracing_position_fetch.cpp [hunk 8]
```diff
@@ -0,0 +1,619 @@
...
+	requested_buffer_device_address_features.bufferDeviceAddress    = VK_TRUE;
+	auto &requested_ray_tracing_features                            = gpu.request_extension_features<VkPhysicalDeviceRayTracingPipelineFeaturesKHR>(VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_RAY_TRACING_PIPELINE_FEATURES_KHR);
+	requested_ray_tracing_features.rayTracingPipeline               = VK_TRUE;
+	auto &requested_acceleration_structure_features                 = gpu.request_extension_features<VkPhysicalDeviceAccelerationStructureFeaturesKHR>(VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_ACCELERATION_STRUCTURE_FEATURES_KHR);
+	requested_acceleration_structure_features.accelerationStructure = VK_TRUE;
+
+	// Sample sepcific feature
+	auto &requested_ray_tracing_position_fetch_features                   = gpu.request_extension_features<VkPhysicalDeviceRayTracingPositionFetchFeaturesKHR>(VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_RAY_TRACING_POSITION_FETCH_FEATURES_KHR);
+	requested_ray_tracing_position_fetch_features.rayTracingPositionFetch = VK_TRUE;
```
Comment (asuessenbach) [inline samples/extensions/ray_tracing_position_fetch/ray_tracing_position_fetch.cpp:75]: This needs to be adjusted when #1013 gets merged.

File: samples/extensions/ray_tracing_position_fetch/ray_tracing_position_fetch.cpp [hunk 9]
```diff
@@ -0,0 +1,619 @@
...
+		    miss_shader_binding_table->get_device_address(),
+		    handle_size_aligned,
+		    handle_size_aligned};
+
+		VkStridedDeviceAddressRegionKHR hit_shader_sbt_entry{
+		    hit_shader_binding_table->get_device_address(),
+		    handle_size_aligned,
+		    handle_size_aligned};
+
+		VkStridedDeviceAddressRegionKHR callable_shader_sbt_entry{};
```
Comment (asuessenbach) [inline samples/extensions/ray_tracing_position_fetch/ray_tracing_position_fetch.cpp]: Aren't those four `VkStridedDeviceAddressRegionKHR` loop invariant? Could then be moved out of the loop.

File: samples/extensions/ray_tracing_position_fetch/ray_tracing_position_fetch.h [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+		glm::mat4 view_inverse{1.0f};
+		glm::mat4 proj_inverse{1.0f};
+		int32_t   display_mode{0};
+	} uniform_data;
+	std::unique_ptr<vkb::core::Buffer> ubo;
+
+	VkPipeline            pipeline{VK_NULL_HANDLE};
+	VkPipelineLayout      pipeline_layout{VK_NULL_HANDLE};
+	VkDescriptorSet       descriptor_set{VK_NULL_HANDLE};
+	VkDescriptorSetLayout descriptor_set_layout{VK_NULL_HANDLE};
```
Comment (asuessenbach) [inline samples/extensions/ray_tracing_position_fetch/ray_tracing_position_fetch.h:62]: Maybe make all the member variables private?

File: samples/extensions/ray_tracing_position_fetch/ray_tracing_position_fetch.h [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+	void         create_shader_binding_tables();
+	void         create_descriptor_sets();
+	void         create_ray_tracing_pipeline();
+	void         create_uniform_buffer();
+	void         build_command_buffers() override;
+	void         update_uniform_buffers();
+	void         draw();
+	virtual void on_update_ui_overlay(vkb::Drawer &drawer) override;
+	bool         prepare(const vkb::ApplicationOptions &options) override;
+	virtual void render(float delta_time) override;
```
Comment (asuessenbach) [inline samples/extensions/ray_tracing_position_fetch/ray_tracing_position_fetch.h:81]: No need to mark `override`ing functions as virtual.

File: framework/core/acceleration_structure.cpp (hunk 2 above)
Comment (SaschaWillems) [reply framework/core/acceleration_structure.cpp]: Nope, that won't work with how the glTF loader works right now. Tried all variants, to no avail.

File: framework/gltf_loader.cpp (hunk 4 above)
Comment (SaschaWillems) [reply framework/gltf_loader.cpp]: Yeah, that was a mistake. Thanks for highlighting this. Will change this the way you suggested above :)

File: framework/gltf_loader.cpp (hunk 3 above)
Comment (SaschaWillems) [reply framework/gltf_loader.cpp]: Done. Thanks for highlighting this 👍🏻

File: samples/extensions/ray_tracing_position_fetch/ray_tracing_position_fetch.cpp (hunk 9 above)
Comment (SaschaWillems) [reply samples/extensions/ray_tracing_position_fetch/ray_tracing_position_fetch.cpp]: Right, I copied that part from other samples and indeed, it makes no sense to have them inside the loop. Also made them const.

File: samples/extensions/ray_tracing_position_fetch/ray_tracing_position_fetch.h (hunk 10 above)
Comment (SaschaWillems) [reply samples/extensions/ray_tracing_position_fetch/ray_tracing_position_fetch.h:62]: We usually only have a public section in all our samples (at least the ones I wrote), so this should be okay.

File: samples/extensions/ray_tracing_position_fetch/ray_tracing_position_fetch.h (hunk 11 above)
Comment (SaschaWillems) [reply samples/extensions/ray_tracing_position_fetch/ray_tracing_position_fetch.h:81]: I copied this from other samples. If this is wrong/should be changed, we need to do it for the other samples too.

File: framework/api_vulkan_sample.cpp (hunk 1 above)
Comment (SaschaWillems) [reply framework/api_vulkan_sample.cpp:1311]: Would love to, but how to set a default value for `vk::BufferUsageFlags` like I do in the non-hpp sample? I don't use the hpp headers personally, so no idea.

Comment (SaschaWillems) [reply framework/api_vulkan_sample.cpp:1311]: I don't have any experience with the hpp headers. How to I convert `vk::BufferUsageFlags` to `VkBufferUsageFlags`?

Comment (asuessenbach) [reply framework/api_vulkan_sample.cpp:1311]: It's not that hard.
Find attached a patch with the port to HPPApiVulkanSample.
[0001-Port-to-HPPApiVulkanSample.patch](https://github.com/user-attachments/files/15882890/0001-Port-to-HPPApiVulkanSample.patch)

File: framework/core/acceleration_structure.cpp (hunk 2 above)
Comment (asuessenbach) [reply framework/core/acceleration_structure.cpp]: Hm... I tried it and it was pretty straight forward. Do I miss anything?
[0001-Use-vkb-core-Buffer-const-in-AccelerationStructure-a.patch](https://github.com/user-attachments/files/15883054/0001-Use-vkb-core-Buffer-const-in-AccelerationStructure-a.patch)

File: framework/api_vulkan_sample.cpp (hunk 1 above)
Comment (SaschaWillems) [reply framework/api_vulkan_sample.cpp:1311]: Awesome. Thank you very much :)

File: framework/core/acceleration_structure.cpp (hunk 2 above)
Comment (SaschaWillems) [reply framework/core/acceleration_structure.cpp]: No, that works totally fine. Probably my lack of experience with C++.

Thank you very much for that patch :)

**Review by gary-sweet — COMMENTED** [review]
Build ok for me and correctly reports not supported. I can't test any more than that sadly.

**Review by asuessenbach — DISMISSED** [review]

**Comment by jeroenbakker-atmind** [comment]
I can test on AMD + Intel GPU this week.

**Comment by jeroenbakker-atmind** [comment]
Works on AMD (RDNA3), Intel doesn't support this extension yet.

image

**Review by jeroenbakker-atmind — APPROVED** [review]

**Review by gpx1000 — DISMISSED** [review]

**Pull request merged**